utils/core/src/serde/mod.rs

```diff
@@ -0,0 +1,146 @@
+use super::{flatten_slice_elements, DeserializationError, Vec};
+
+mod byte_reader;
+pub use byte_reader::{ByteReader, SliceReader};
+
+mod byte_writer;
+pub use byte_writer::ByteWriter;
+
+// SERIALIZABLE TRAIT
+// ================================================================================================
+
+/// Defines how to serialize `Self` into bytes.
+pub trait Serializable: Sized {
+    // REQUIRED METHODS
+    // --------------------------------------------------------------------------------------------
+    /// Serializes `self` into bytes and writes these bytes into the `target`.
+    fn write_into<W: ByteWriter>(&self, target: &mut W);
+
+    // PROVIDED METHODS
+    // --------------------------------------------------------------------------------------------
+
+    /// Serializes `self` into a vector of bytes.
+    fn to_bytes(&self) -> Vec<u8> {
+        let mut result = Vec::with_capacity(self.get_size_hint());
+        self.write_into(&mut result);
+        result
+    }
+
+    /// Serializes all elements of the `source` and writes these bytes into the `target`.
+    ///
+    /// This method does not write any metadata (e.g. number of serialized elements) into the
+    /// `target`.
+    fn write_batch_into<W: ByteWriter>(source: &[Self], target: &mut W) {
```

Comment:
We could make this more flexible by taking an iterator as argument instead of a slice.

```rust
fn write_batch_into<T, I, W>(source: I, target: &mut W) where
T: core::borrow::Borrow<Self>,
I: IntoIterator<Item = T>,
W: ByteWriter
```

This way, we will allow anything to be used as argument - including slices.

Reply:
I like this, and tried to make it work, but ran into lifetime bounds issues - so, leaving this for the future.